mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java

```diff
@@ -96,6 +106,23 @@
 
     @Override
     public Task deleteExpiredMessagesTask() {
-        throw new NotImplementedException("Will be implemented later");
+        ZonedDateTime now = ZonedDateTime.now(clock);
+        ZonedDateTime beginningOfRetentionPeriod = now.minus(retentionConfiguration.getRetentionPeriod());
+
+        Flux<BucketName> deleteOperation = retentionQualifiedBuckets(beginningOfRetentionPeriod)
+            .flatMap(bucketName -> deleteBucketData(bucketName).then(Mono.just(bucketName)));
+
+        return new BlobStoreVaultGarbageCollectionTask(beginningOfRetentionPeriod, deleteOperation);
+    }
+
+    @VisibleForTesting
+    Flux<BucketName> retentionQualifiedBuckets(ZonedDateTime beginningOfRetentionPeriod) {
+        return Flux.from(messageMetadataVault.listRelatedBuckets())
+            .filter(bucketName -> nameGenerator.isBucketRangeBefore(bucketName, beginningOfRetentionPeriod));
```

Comment:
In bucket `dmsg-2018-07-01` we have email ranging from 01/07/2018 to 31/07/2018.

We are 10/07/2018. I apply retention. You are using the starting date for the 1 year retention. Thus I will delete mails of the 23/07/2019. Which is not already 1 year.

We should use the ending date associated with the bucket, not the starting one. Carefull about month being 28-29-30-31 days when doing the maths...

Reply:
I would prefer a 

```
buckets.filter(bucket -> generator.endDate(bucket).isBefore(retentionStart));
```

Reply:
What happens when the name pattern does not match? (we should ERROR log + not fail)